docs/src/content/docs/patterns/structured-outputs.mdx

```diff
@@ -0,0 +1,155 @@
+
+First, define the input and output types for the component.
+
+```tsx
+interface ExtractEntitiesProps {
+  text: string;
+}
+
+interface ExtractEntitiesOutput {
+  people: string[];
+  places: string[];
+  organizations: string[];
+}
+```
+
+Next, define the component and use the `ChatCompletion` component to get the structured output.
+
+```tsx
+const ExtractEntities = gsx.Component<
+  ExtractEntitiesProps,
+  ExtractEntitiesOutput
+>(async ({ text }) => {
+  const prompt = `Please review the following text and extract all the people, places, and organizations mentioned.
+
+  <text>
+  ${text}
+  </text>
+    
+  Please return json with the following format:
+  {
+    "people": ["person1", "person2", "person3"],
+    "places": ["place1", "place2", "place3"],
+    "organizations": ["org1", "org2", "org3"]
+  }`;
+  return (
+    <ChatCompletion
+      model="gpt-4o-mini"
+      messages={[
+        {
+          role: "user",
+          content: prompt,
+        },
+      ]}
+      response_format={{ type: "json_object" }}
+    >
+      {(response: string) => {
+        return JSON.parse(response) as ExtractEntitiesOutput;
+      }}
+    </ChatCompletion>
+  );
+});
+```
+
+Using `response_format={{ type: "json_object" }}` ensure that we get valid JSON back, but it doesn't guarantee that the JSON will match the `ExtractEntitiesOutput` type. Parsing this way also doesn't provide any type safety.
+
+## Using Zod
+
+You can can take this one step further by using Zod to supply a JSON schema which guarantees that the model response will match the `ExtractEntitiesOutput` type.
+
+```tsx
+import { z } from "zod";
+import { zodResponseFormat } from "openai/helpers/zod";
+
+// Define the Zod schema
+const ExtractEntitiesSchema = z.object({
+  people: z.array(z.string()),
+  places: z.array(z.string()),
+  organizations: z.array(z.string()),
+});
+
+type ExtractEntitiesOutput = z.infer<typeof ExtractEntitiesSchema>;
+```
+
+After definining this, you just need to make a few small changes to the component to use the Zod schema.
+
+```tsx
+const ExtractEntities = gsx.Component<
+  ExtractEntitiesProps,
+  ExtractEntitiesOutput
+>(async ({ text }) => {
+  const prompt = `Please review the following text and extract all the people, places, and organizations mentioned.
+
+  <text>
+  ${text}
+  </text>
+    
+  Please return json with the following format:
+  {
+    "people": ["person1", "person2", "person3"],
+    "places": ["place1", "place2", "place3"],
+    "organizations": ["org1", "org2", "org3"]
+  }`;
+  return (
+    <ChatCompletion
+      model="gpt-4o-mini"
+      messages={[
+        {
+          role: "user",
+          content: prompt,
+        },
+      ]}
+      response_format={zodResponseFormat(ExtractEntitiesSchema, "entities")}
+    >
+      {(response: string) => {
+        try {
+          return ExtractEntitiesSchema.parse(JSON.parse(response));
```

Comment:
I assume you ran this code and it works, but just want to confirm? 

Would it be worth turning this into a complete example that is checked in that we can link to from this doc?

Reply:
Yep, i tested it (which was good because it was wrong before i tested it :))

> Would it be worth turning this into a complete example that is checked in that we can link to from this doc?

Good question. A full example with a package.json and everything feels like overkill but maybe we can have somewhere where we have the full code examples from the docs. might be worth chatting about more tomorrow.

Reply:
I'm envisioning a `basicExamples` folder within examples where each example from docs is it's own tsx file. then could wire up a pnpm command for each if we wanted. Might mess up the current test setup.